pkg/validator/shoot_validator.go

```diff
@@ -0,0 +1,159 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validator
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	gcpvalidation "github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp/validation"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/gardener/gardener/pkg/apis/core"
+)
+
+var (
+	specPath = field.NewPath("spec")
+
+	networkPath  = specPath.Child("networking")
+	providerPath = specPath.Child("provider")
+
+	infrastructureConfigPath = providerPath.Child("infrastructureConfig")
+	controlPlaneConfigPath   = providerPath.Child("controlPlaneConfig")
+	workersPath              = providerPath.Child("workers")
+)
+
+type validationContext struct {
+	shoot                *core.Shoot
+	infrastructureConfig *gcp.InfrastructureConfig
+	controlPlaneConfig   *gcp.ControlPlaneConfig
+	cloudProfile         *gardencorev1beta1.CloudProfile
+	cloudProfileConfig   *gcp.CloudProfileConfig
+}
+
+// getAllowedRegionZonesFromCloudProfile fetches the set of allowed zones from the Cloud Profile.
+func getAllowedRegionZonesFromCloudProfile(shoot *core.Shoot, cloudProfile *gardencorev1beta1.CloudProfile) sets.String {
+	shootRegion := shoot.Spec.Region
+	for _, region := range cloudProfile.Spec.Regions {
+		if region.Name == shootRegion {
+			gcpZones := sets.NewString()
+			for _, gcpZone := range region.Zones {
+				gcpZones.Insert(gcpZone.Name)
+			}
+			return gcpZones
+		}
+	}
+	return nil
+}
+
+func (v *Shoot) validateShoot(valContext *validationContext) field.ErrorList {
+	var (
+		allErrors    = field.ErrorList{}
+		allowedZones = getAllowedRegionZonesFromCloudProfile(valContext.shoot, valContext.cloudProfile)
+	)
+
+	allErrors = append(allErrors, gcpvalidation.ValidateNetworking(valContext.shoot.Spec.Networking, networkPath)...)
+	allErrors = append(allErrors, gcpvalidation.ValidateInfrastructureConfig(valContext.infrastructureConfig, valContext.shoot.Spec.Networking.Nodes, valContext.shoot.Spec.Networking.Pods, valContext.shoot.Spec.Networking.Services, infrastructureConfigPath)...)
+	allErrors = append(allErrors, gcpvalidation.ValidateControlPlaneConfig(valContext.controlPlaneConfig, allowedZones, controlPlaneConfigPath)...)
```

Comment:
I remember a discussion together with @rfranzke in which we agreed that the zone given in the `ControlPlaneConfig` must at leat be used by one worker pool. Isn't that still relevant?

Reply:
I remember the discussion, and probably I have the incorporated the outcome in the older PR which was closed in favour of this one (I just copied what I did there here). In any-case, I added this case to the validation.